src/core/ddsi/src/q_transmit.c

```diff
@@ -1024,6 +1068,117 @@ static int maybe_grow_whc (struct writer *wr)
   return 0;
 }
 
+int write_sample_p2p_wrlock_held(struct writer *wr, seqno_t seq, struct nn_plist *plist, struct ddsi_serdata *serdata, struct ddsi_tkmap_instance *tk, struct proxy_reader *prd)
+{
+  struct q_globals * const gv = wr->e.gv;
+  int r;
+  nn_mtime_t tnow;
+  int rexmit = 1;
+  struct wr_prd_match *wprd = NULL;
+  seqno_t gseq;
+  struct nn_xmsg *gap = NULL;
+  seqno_t gapstart = -1, gapend = -1, segstart;
+  unsigned gapbits[256 / 32];
+  unsigned gapnumbits = 0;
+
+  tnow = now_mt ();
+  serdata->twrite = tnow;
+  serdata->timestamp = now();
+
+  if (prd->filter)
+  {
+    if ((wprd = ddsrt_avl_lookup (&wr_readers_treedef, &wr->readers, &prd->e.guid)) != NULL)
+    {
+      rexmit = prd->filter(wr, prd, serdata);
+      /* determine if gap has to added */
+      if (rexmit)
+      {
+        GVLOG (DDS_LC_DISCOVERY, "send filtered "PGUIDFMT" last_seq=%"PRIu64" seq=%"PRIu64"\n", PGUID (wr->e.guid), wprd->seq, seq);
+
+        segstart = wprd->seq + 1;
+        memset (gapbits, 0, sizeof (gapbits));
+        for (gseq = segstart; gseq < seq; gseq++)
+        {
+          struct whc_borrowed_sample sample;
+          if (whc_borrow_sample (wr->whc, seq, &sample))
+          {
+            if (prd->filter(wr, prd, sample.serdata) == 0)
+            {
+              if (gapstart == -1)
+              {
+                GVLOG (DDS_LC_DISCOVERY, " M%"PRId64, gseq);
+                gapstart = gseq;
+                gapend = gapstart + 1;
+              }
+              else if (gseq == gapend)
+              {
+                GVLOG (DDS_LC_DISCOVERY, " M%"PRId64, gseq);
+                gapend = gseq + 1;
+              }
+              else if (gseq - gapend < 256)
+              {
+                unsigned idx = (unsigned) (gseq - gapend);
+                GVLOG (DDS_LC_DISCOVERY, " M%"PRId64, gseq);
+                gapnumbits = idx + 1;
+                nn_bitset_set (gapnumbits, gapbits, idx);
+              }
+            }
+            whc_return_sample (wr->whc, &sample, false);
+          }
+        }
+
+        if (gapstart > 0)
+        {
+          if (gapnumbits == 0)
+          {
+            /* Avoid sending an invalid bitset */
```

Comment:
DDSI 2.3 has made it legal to send an empty bitset. Cyclone doesn't yet declare itself to be DDSI 2.3 (though it pretty much is), and it makes sense to get rid of these crazy workarounds for what have been declared to be errors in the spec. (``handle_NackFrag`` has the same problem though ...)